### DISCUSSION TITLE
Use SMB shares for storing uploaded/converted files

### DISCUSSION THREAD

**Discussion by illuminati05** [mediacms-io/mediacms#530] [Q&A]
**Describe the feature you'd like**
A possibilty to change the working directories in the settings.py from local directories to SMB shares
My SMB shares are huge 32TB. And my xcp-ng VM machines has only 10GB with Ubuntu Server. I want to keep the VM at small size, because of backups.

The reason: I have already a big amount of videos,photos and so on. When i upload these files to mediacms i have them on the VM and on my TRUENAS storage. 

**Describe alternatives you've considered**
-

**Additional context**
I tried to mount SMB shares under /media/"FOLDER" with success, but in the settings.py i dont know which entries must be modified to use the SMB shares for (video decoding/MEDIA_ROOT)

**Comment by mgogoulos** [comment]
mediacms_files is what interests you, that's the directory that holds all media (original and encoded/transcoded versions). 
If you can mount it as a disk/filesystem/partition, this will work for you. 
So something like this would work I assume (haven't tried it with SMB but have tried with mounting a different disk on mediacms_files):

a. install mediacms somewhere, eg /home/user/mediacms
b. mount SMB share on /home/user/mediacms/mediacms_files 
c. now when you upload something it goes there. If you unmount the SMB share it won't be available (because it 'lives' on the SMB share)

Hope this helps

**Reply by LianhuanJ** [reply]
Might the path _mediacms_files_ mentioned above be **media_files** ? which is misleading

**Reply by supraxd2021** [reply]
how to mount?

**Comment by illuminati05** [comment]
So i have to create all the subfolders like on the local folder? Or will they be created automatically after encoding?
Edit the fstab file and make a link for the "media_files" to "//server/share/"?

**Reply by mgogoulos** [reply]
they will be created automatically! But mediacms should have the permissions to write there!

**Comment by marcofloriano** [comment]
**b. mount SMB share on /home/user/mediacms/mediacms_files**

Could you please give some example/tutorial on how to do that on Ubuntu 20 ? Thanks

**Comment by Nanodk** [comment]
Sorry to hi-jack and rebump this tread, but I'm also struggling to get this to work. 

Full access to network share (SMB on TrueNAS) have copyed all from media_files to same share. can't upload, but viewing is working.

When I try to upload it show x% and then retrying 2 times and fail. 

my Fstab line:
//path/to/Media/mediacms  /home/mediacms.io/mediacms/m_files  cifs  guest,iocharset=utf8,file_mode=0777,dir_mode=0777,credentials=/home/mediacms.io/smbcredentials.cfg

**Reply by MAXimized490** [reply]
I'm having this exact issue. Upload will stop quickly after it starts, retry twice, and fail. I am using Docker Compose. All it seems to do is make an hls folder. Nothing else.

Edit: While I am still having other issues with MediaCMS, I was at least able to resolve this one and it was simply due to a lack of knowledge with fstab and SMB on Linux. The issue is that the share was being mounted in a way where my user's uid and gid were being used for ownership on all files within the share. The www-data user needs to have access to these files, so I made the following changes in my fstab entry:

`//<server> <path-to-mediacms>/media_files cifs credentials=<credentials-file>,uid=33,gid=1000,file_mode=0775,dir_mode=0775 0 0`

The key things here are in the ",uid=33,gid=1000,file_mode=0775,dir_mode=0775". For me, www-data user was 33 and 1000 is my user's group. I left read access for everyone because I wasn't sure if MediaCMS was accessing these areas with other users, and it seemed liked www-data was the only one writing here.

**Comment by Nanodk** [comment]
Neverminde, made at NFS share, mounted it as medie_files, made www-data RW on it and auto mount white fstab, works now :)

**Comment by xansone** [comment]
> Neverminde, made at NFS share, mounted it as medie_files, made www-data RW on it and auto mount white fstab, works now :)

Hey, would you be so kind and explain excatly how did you do it? :) I have permission problem, excatly like you :)

**Comment by XLUGO74** [comment]
I am getting:

PermissionError : [Errno 13] Permission denied : 'my_path'

when I change the default directory

**Comment by Bnewn** [comment]
Also need help with this! MediaCMS is able to launch and detect the smb mounted share properly and create it's media_files folder with hls folder within it. However, when I go to upload a video it keeps on failing.

I tried changing the bind to be on the same directory as the compose file and it works fine. Any ideas? Thank you in advance

**Reply by dustojnikhummer** [reply]
> Also need help with this! MediaCMS is able to launch and detect the smb mounted share properly and create it's media_files folder with hls folder within it. However, when I go to upload a video it keeps on failing.
> 
> I tried changing the bind to be on the same directory as the compose file and it works fine. Any ideas? Thank you in advance

Sounds to me like a permission issue. Can you write files as the CMS user? Docker install?

**Comment by mrjoshuap** [comment]
First, TLDR: unless you have a solid understanding of Unix UID and GID and know the impacts of container runtimes on those permissions, you should use NFS.

That said, SMB/CIFS for media files does work and is more performant with a big caveat that you need to authenticate and permissions are more complex vs NFS.

This affects both container environments and regular deployments. Ultimately you have to figure out how you potentially mount with one user and have read/write to another user or two...

Just make sure you don't run your database on a network share  🤣
